nav2_mppi_controller/src/trajectory_visualizer.cpp

```diff
@@ -104,27 +107,95 @@ void TrajectoryVisualizer::add(
 }
 
 void TrajectoryVisualizer::add(
-  const models::Trajectories & trajectories, const std::string & marker_namespace)
+  const models::Trajectories & trajectories, const Eigen::ArrayXf & costs,
+  const std::string & marker_namespace,
+  const builtin_interfaces::msg::Time & cmd_stamp)
 {
   size_t n_rows = trajectories.x.rows();
   size_t n_cols = trajectories.x.cols();
-  const float shape_1 = static_cast<float>(n_cols);
-  points_->markers.reserve(floor(n_rows / trajectory_step_) * floor(n_cols * time_step_));
+  points_->markers.reserve(n_rows / trajectory_step_);
 
-  for (size_t i = 0; i < n_rows; i += trajectory_step_) {
-    for (size_t j = 0; j < n_cols; j += time_step_) {
-      const float j_flt = static_cast<float>(j);
-      float blue_component = 1.0f - j_flt / shape_1;
-      float green_component = j_flt / shape_1;
+  // Use percentile-based normalization to handle outliers
+  // Sort costs to find percentiles
+  std::vector<float> sorted_costs(costs.data(), costs.data() + costs.size());
+  std::sort(sorted_costs.begin(), sorted_costs.end());
 
-      auto pose = utils::createPose(trajectories.x(i, j), trajectories.y(i, j), 0.03);
-      auto scale = utils::createScale(0.03, 0.03, 0.03);
-      auto color = utils::createColor(0, green_component, blue_component, 1);
-      auto marker = utils::createMarker(
-        marker_id_++, pose, scale, color, frame_id_, marker_namespace);
+  // Use 10th and 90th percentile for robust color mapping
```

Comment:
I found this percentile separation to be the best for my case

Reply:
If we do the eigen normalization, we can promise normalization from 0-1 always. I think that would be the best?